cni/pkg/ambient/informers_test.go

```diff
@@ -0,0 +1,283 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ambient
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/test/util/assert"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+const (
+	testNamespace = "test-ns"
+)
+
+func TestServerReconcilePod(t *testing.T) {
+	ctx := context.Background()
+
+	setup := func(ctx context.Context) (kube.CLIClient, *fakePodReconcileHandler, error) {
+		kubeClient := kube.NewFakeClient()
+		server := &Server{
+			ctx:        ctx,
+			kubeClient: kubeClient,
+		}
+		fakePodHandler := &fakePodReconcileHandler{}
+		server.podReconcileHandler = fakePodHandler
+		server.setupHandlers()
+		server.Start()
+
+		_, err := kubeClient.Kube().CoreV1().Namespaces().Create(ctx, &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   testNamespace,
+				Labels: map[string]string{"istio.io/dataplane-mode": "ambient"},
+			},
+		}, metav1.CreateOptions{})
+
+		return kubeClient, fakePodHandler, err
+	}
+
+	type expectedEvent struct {
+		event     string
+		podName   string
+		assertPod func(pod *v1.Pod) bool
+		// couldSkip is used to indicate that the event could be skipped in assertion.
+		couldSkip bool
+	}
+
+	cases := []struct {
+		name           string
+		podOperations  func(ctx context.Context, kubeClient kube.CLIClient) error
+		expectedEvents []expectedEvent
+	}{
+		{
+			name: "add a pod to mesh",
+			podOperations: func(ctx context.Context, kubeClient kube.CLIClient) error {
+				_, err := kubeClient.Kube().CoreV1().Pods(testNamespace).Create(ctx, &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-1",
+						Namespace: testNamespace,
+					},
+				}, metav1.CreateOptions{})
+				time.Sleep(1 * time.Second)
```

Comment:
```suggestion
```

Probably not required  - and if it is and there's a good reason for it, better to have a recheck/timeout loop around the actual assert step.